### PR TITLE
typescript, promises, oh my!

### PR DESCRIPTION
**Pull request by JustinBeckwith**
image

This does a few things:
- Converts the module to TypeScript
- Eliminates the `async` dependency
- Converts the interface over to Promises instead of callbackies

It's waiting for a release of `is-stream-end` before we can merge. 

Feedback welcomed!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,11 +27,15 @@
...
+    "@types/through2": "^2.0.33",
+    "gts": "^0.5.4",
+    "mocha": "^5.0.5",
+    "through2": "^2.0.3",
+    "typescript": "~2.8.1"
   },
   "dependencies": {
-    "async": "^2.4.0",
-    "is-stream-ended": "^0.1.0"
+    "is-stream-ended": "file:../is-stream-ended/is-stream-ended-0.1.3.tgz"
```
Comment (stephenplusplus) [inline package.json]: Published as 0.1.4

**Review by stephenplusplus — CHANGES_REQUESTED** [review]
Could you kill package-lock.json from this PR? I'm still not a big fan :(-- hoping it gets easier to work with as time goes, but it's still a big pain point for GC* at this time (and according to Google + StackOverflow, many others).

Feel free to add it to the new .gitignore :)

File: package.json (hunk 1 above)
Comment (JustinBeckwith) [reply package.json]: Thank you!

**Comment by JustinBeckwith** [comment]
Ooops, didn't even notice that got swooped up with the other changes.  Officially ignored :)

**Comment by stephenplusplus** [comment]
Could you update the README as well?

**Review by stephenplusplus — APPROVED** [review]

**Comment by stephenplusplus** [comment]
Thanks!

**Comment by stephenplusplus** [comment]
Published as 2️⃣💠0️⃣💠0️⃣

**Pull request merged**